serverless/cmd/feeder/github/main.go

```diff
@@ -0,0 +1,561 @@
+	}
+	glog.V(1).Infof("Cloned %q into %q", cloneOpts.URL, dest)
+
+	// Create a remote -> logOwnerRepo
+	//  git remote add upstream "https://github.com/${log_repo}.git"
+	logURL := fmt.Sprintf("https://github.com/%v.git", opts.logRepo)
+	logRemote, err := forkRepo.CreateRemote(&config.RemoteConfig{
+		Name: "upstream",
+		URLs: []string{logURL},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to add remote %q to fork repo: %v", logURL, err)
+	}
+	glog.V(1).Infof("Added remote upstream->%q for %q:\n%v", logURL, opts.witnessRepo, logRemote)
+
+	//  git fetch --all
+	if err = forkRepo.FetchContext(ctx, &git.FetchOptions{}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return nil, fmt.Errorf("failed to fetch repo %q: %v", *opts.witnessRepo, err)
+	}
+
+	return forkRepo, nil
+}
+
+func authWithGithub(ctx context.Context, ghToken string) (*github.Client, error) {
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: ghToken})
+	tc := oauth2.NewClient(ctx, ts)
+	return github.NewClient(tc), nil
+}
+
+func pullAndGetRepoHead(r *git.Repository) (*plumbing.Reference, *git.Worktree, error) {
+	// Figure out where the worktree is for repo 'r'.
+	wt, err := r.Worktree()
+	if err != nil {
+		return nil, nil, fmt.Errorf("workTree(%v) err: %v", r, err)
+	}
+	// Pull the latest commits from remote 'origin'.
+	if err := wt.Pull(&git.PullOptions{RemoteName: "origin"}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return nil, nil, fmt.Errorf("git pull %v/origin err: %v", r, err)
+	}
+	// Get the master HEAD - we'll need this to branch from later on.
+	headRef, err := r.Head()
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %v HEAD ref err: %v", r, err)
+	}
+	// Print the latest commit that was just pulled
+	commit, err := r.CommitObject(headRef.Hash())
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %v HEAD commit err: %v", r, err)
+	}
+	glog.V(1).Infof("Reading %v: HEAD commit:\n%v", r, commit)
+	return headRef, wt, nil
+}
+
+// feedOnce performs a one-shot "feed to witness and create PR" operation.
+func feedOnce(ctx context.Context, opts *options, forkRepo *git.Repository, ghCli *github.Client) error {
+	// Pull forkRepo and get the ref for origin/master branch HEAD.
+	headRef, workTree, err := pullAndGetRepoHead(forkRepo)
+	if err != nil {
+		return err
+	}
+
+	// TODO: if checkpoint and checkpoint.witnessed bodies are different (strictly, if checkpoint is newer),
+	// then we need to be feeding checkpoint to the witness, otherwise we can feed the witnessed one and
+	// short-circuit creating a PR if our witness(es) has/have already signed it.
+	inputCP := filepath.Join(opts.witnessClonePath, opts.logPath, "checkpoint")
+	glog.V(1).Infof("Reading CP from %q", inputCP)
+	cp, err := util.ReadFile(workTree.Filesystem, inputCP)
+	if err != nil {
+		return fmt.Errorf("failed to read input checkpoint: %v", err)
+	}
+
+	// 2. kick off feeder.
```

Comment:
This number has got lost along the way. We have 2, 5, 6, 7, 8 in this method.

I tried to tie it back to the intro to the main loop, but it claims `7. --end of cycle--` so then I got even more confused.

Reply:
hehe, I'll remove them